### PR TITLE
Improve code lens generation time

### PR DESCRIPTION
**Pull request by kpbochenek**
Currently when we search for a symbol definition we firstly check if it exists in current document.
We do that by going one by one and comparing. It is okay for single pass, but checking multiple symbols is very inefficient. This converts list of symbols to a map [symbol -> symbolinformation].

When we found parent class for given method we fetched each method of parent class and checked if it is super method. We compared method 'f' to X#aaa(), X#bbb(), X#f(), but it is easy to see don't need to fetch methods and compare types if names don't match.

Added a map to memoize SymbolInformation for single request. I will verify how much it actually helps but wanted to get your opinion what you think.

Added logging if lenses generation takes longer than 1s.

### CODE REVIEW

**Comment by olafurpg** [comment]
I agree with @tgodzik I think it's a good suggestion to setup a JMH benchmark for `Global.scala`. Benchmarks were essential when I was working on performance of completions https://github.com/scalameta/metals/blob/ef8d081b370a476a99cee1f826c5d1cc26bb1dfa/metals-bench/src/main/scala/bench/CompletionBench.scala

File: metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala [hunk 1]
```diff
@@ -88,6 +88,25 @@ final class ImplementationProvider(
       textDocumentWithPath.textDocument
     )
 
+  def defaultSymbolSearchMemoize(
+      anyWorkspacePath: AbsolutePath,
+      textDocument: TextDocument
+  ): String => Option[SymbolInformation] = {
+    lazy val global =
```
[the remark this answers is not in the document]
Comment (kpbochenek) [reply metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala:95]: good catch! as globalTable is available in ImplementationProvider I changed this to:
`globalTable.globalSymbolTableFor(anyWorkspacePath)`

File: metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala [hunk 2]
```diff
@@ -197,6 +197,7 @@ object SuperMethodProvider {
         documentWithPath.textDocument,
         findSymbol
       )
+    val methodName = msi.symbol.desc.name.value
```
[the remark this answers is not in the document]
Comment (kpbochenek) [reply metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala]: changed

**Comment by kpbochenek** [comment]
cool I will take a look at JMH benchmark but right now I need to focus back on dotty, would it be okay to merge this PR as it significantly improve lenses generation and create another issue to specifically setup JMH benchmark and look for more performance problems?

**Comment by olafurpg** [comment]
Sounds good to me 👍 this PR is already a big improvement!

**Pull request merged**